build/xtask/src/dist.rs

```diff
@@ -1821,105 +1864,50 @@ pub fn allocate_all(
         let mut free = toml.memories(image_name)?;
         let kernel_requests = &kernel.requires;
 
-        let mut task_requests: BTreeMap<&str, BTreeMap<u32, VecDeque<&str>>> =
+        let mut task_requests: BTreeMap<&str, IndexMap<&str, VecDeque<u32>>> =
             BTreeMap::new();
 
         for name in tasks.keys() {
-            for (mem, amt) in task_sizes[name.as_str()].iter() {
-                let bytes = toml.suggest_memory_region_size(name, *amt);
+            let req = &task_sizes[name.as_str()];
+            for (&mem, &amt) in req.memory.iter() {
+                // Right now, flash is most limited, so it gets to use all of
+                // our spare regions (if present)
+                let n = if mem == "flash" {
+                    req.spare_regions + 1
+                } else {
+                    1
+                };
+                let bytes = toml.suggest_memory_region_size(name, amt, n);
                 if let Some(r) = tasks[name].max_sizes.get(&mem.to_string()) {
-                    if bytes > *r as u64 {
+                    let total_bytes = bytes.iter().sum::<u64>();
+                    if total_bytes > *r as u64 {
```

Comment:
nit: this looks like a lossy conversion but isn't, consider `u64::from(*r)`

Reply:
Good catch, done in 3e2421a76e45b021398152745b9c6ebb2e871d01